File: algorithm-cpp/c03_insertion_sort.cpp
#include<iostream>
#include<vector>
using namespace std;

void insertionSort(vector<int> &arr);
void swap(vector<int> &arr, int i, int j);

int main()
{
    vector<int> vec = {5, 3, 7, 2, 9};
    insertionSort(vec);
    for (int i=0; i<vec.size(); i++) {
        cout << vec[i] << " ";
    }
    cout << endl;
}

void insertionSort(vector<int> &arr)
{
    for (int i=1; i<arr.size(); i++)
    {
        int index = i;
        while (index > 0 && arr[index] < arr[index-1])
        {
            swap(arr, index, index - 1);
            index--;
        }
    }
}

void swap(vector<int> &arr, int i, int j)
{
    if (i == j)
    {
        return;
    }
    arr[i] = arr[i] ^ arr[j];
    arr[j] = arr[i] ^ arr[j];
    arr[i] = arr[i] ^ arr[j];
}

File: algorithm-java/src/com/godric/v21_11/algo/C03_InsertionSort.java
package com.godric.v21_11.algo;

import com.godric.v21_11.utils.ArrayUtil;

public class C03_InsertionSort {

    public static void insertionSort(int[] arr) {
        if (arr == null || arr.length < 2) {
            return;
        }
        for (int i=1; i<arr.length; i++) {
            int index = i;
            while (index > 0 && arr[index] < arr[index - 1]) {
                ArrayUtil.swap(arr, index, index - 1);
                index--;
            }
        }
    }

    public static void main(String[] args) {
        int maxSize = 5000;
        int maxValue = 10000;
        int testTimes = 50000;

        while (testTimes-- > 0) {
            int[] arr = ArrayUtil.generateRandomArr(maxSize, maxValue);
            int[] arr1 = ArrayUtil.copyArr(arr);
            int[] arr2 = ArrayUtil.copyArr(arr);

            insertionSort(arr1);
            ArrayUtil.comparator(arr2);

            if (!ArrayUtil.isEquals(arr1, arr2)) {
                System.out.println("You have wrong in selection sort, arr :");
                ArrayUtil.printArr(arr);
                System.out.println("insertion sort result:");
                ArrayUtil.printArr(arr1);
                System.out.println("system sort result:");
                ArrayUtil.printArr(arr2);
                return;
            }
            if (testTimes % 100 == 0) {
                System.out.println("wait please, lost time is " + testTimes);
            }
        }

        System.out.println("nice！");
    }

}
